riscv-rt/build.rs

```diff
@@ -11,6 +11,15 @@ fn add_linker_script(arch_width: u32) -> io::Result<()> {
     let mut content = fs::read_to_string("link.x.in")?;
     content = content.replace("${ARCH_WIDTH}", &arch_width.to_string());
 
+    // Single core devices can use the entire stack, no need to limit it.
+    let hart_stack_size = if env::var_os("CARGO_FEATURE_SINGLE_HART").is_some() {
+        "SIZEOF(.stack)"
+    } else {
+        "2K"
```

Comment:
Would something like this make sense?

```suggestion
        "SIZEOF(.stack) / (_max_hart_id + 1)"
```

Reply:
Agreed, the environment variable check can probably be skipped.

Reply:
I think I agree too. I have a question though since I don't have much experience with multi core devices.

If we default to splitting the stack equal parts among the cores (4 in case of 4 cores), do we need to add some more checks?

I see there's a heap variable with a default size of 0. If someone have currently set the heap to something that will cause the stack of their last core to become smaller than the others.

Is this something that needs to be taken into account?

Reply:
Ahh, yes i see, because .heap does ` . += _heap_size;`

With heap set to 1024:
.heap                      0  0x1000008
.stack                  4088  0x1000008

With heap set to 1024:
.heap                   1024  0x1000008
.stack                  3064  0x1000408

I checked with `nm <elf> | grep _hart_stack_size` and it looks like it works even with sizes not perfectly divisable. But I'm unsure about the alignment. For example with 3 harts I get a stack size of 1362.

It looks like these instructions in _start might be there to fix that that though?

```
cfg_global_asm!(
    "la t1, _stack_start",
    #[cfg(not(feature = "single-hart"))]
    "sub t1, t1, t0",
    "andi sp, t1, -16 // align stack to 16-bytes
    add s0, sp, zero",
);
```

This level of assembler and bit fibbling is not my strong suite.

Reply:
> But I'm unsure about the alignment. For example with 3 harts I get a stack size of 1362.

We can probably include an assert to ensure that the resulting stack size(s) satisfy alignment requirements.

> It looks like these instructions in _start might be there to fix that that though?

Yes, but as I understand it, the stack offsets will also need to be properly aligned. So, having the assert to check at the linking stage will still be useful.